File: advent-of-code/2015/05/02/solution.py
import os

def is_nice(string):
    # Check for a pair that appears at least twice without overlapping
    has_pair = False
    for i in range(len(string) - 1):
        pair = string[i:i+2]
        # Look for the same pair starting at least 2 positions later
        if pair in string[i+2:]:
            has_pair = True
            break
    
    if not has_pair:
        return False
    
    # Check for a letter that repeats with exactly one letter between
    has_repeat = any(string[i] == string[i+2] for i in range(len(string)-2))
    
    return has_repeat

# Test cases with assertions
assert is_nice("qjhvhtzxzqqjkmpb") == True
assert is_nice("xxyxx") == True
assert is_nice("uurcxstgmygtbstg") == False
assert is_nice("ieodomkazucvgmuy") == False

# Read input and calculate result
script_dir = os.path.dirname(os.path.abspath(__file__))
input_path = os.path.join(script_dir, "..", "input.txt")
with open(input_path) as f:
    strings = f.read().strip().split('\n')

result = sum(1 for s in strings if is_nice(s))
print(f"Python: {result}")

File: advent-of-code/2015/05/02/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func isNice(s string) bool {
	// Check for a pair that appears at least twice without overlapping
	hasPair := false
	for i := 0; i < len(s)-1; i++ {
		pair := s[i : i+2]
		// Look for the same pair starting at least 2 positions later
		if strings.Contains(s[i+2:], pair) {
			hasPair = true
			break
		}
	}
	if !hasPair {
		return false
	}

	// Check for a letter that repeats with exactly one letter between
	hasRepeat := false
	for i := 0; i < len(s)-2; i++ {
		if s[i] == s[i+2] {
			hasRepeat = true
			break
		}
	}

	return hasRepeat
}

func main() {
	// Test cases with assertions
	if !isNice("qjhvhtzxzqqjkmpb") {
		panic("Test failed: isNice(\"qjhvhtzxzqqjkmpb\") != true")
	}
	if !isNice("xxyxx") {
		panic("Test failed: isNice(\"xxyxx\") != true")
	}
	if isNice("uurcxstgmygtbstg") {
		panic("Test failed: isNice(\"uurcxstgmygtbstg\") != false")
	}
	if isNice("ieodomkazucvgmuy") {
		panic("Test failed: isNice(\"ieodomkazucvgmuy\") != false")
	}

	// Read input and calculate result
	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)
	inputPath := filepath.Join(dir, "..", "input.txt")
	content, err := os.ReadFile(inputPath)
	if err != nil {
		panic(err)
	}
	strings_list := strings.Split(strings.TrimSpace(string(content)), "\n")

	count := 0
	for _, s := range strings_list {
		if isNice(s) {
			count++
		}
	}

	fmt.Printf("Go: %d\n", count)
}
